container_api/src/container/ui_admin.rs

```diff
@@ -0,0 +1,389 @@
+            .ui_interfaces
+            .clone()
+            .into_iter()
+            .filter(|ui_interface| ui_interface.bundle == id.to_string());
+
+        for bundle_interface in to_remove {
+            self.remove_ui_interface(&bundle_interface.id)?;
+        }
+
+        new_config.check_consistency()?;
+        self.config = new_config;
+        self.save_config()?;
+        Ok(())
+    }
+
+    fn add_ui_interface(
+        &mut self,
+        new_interface: UiInterfaceConfiguration,
+    ) -> Result<(), HolochainError> {
+        let mut new_config = self.config.clone();
+        new_config.ui_interfaces.push(new_interface.clone());
+        new_config.check_consistency()?;
+        self.config = new_config;
+        self.save_config()?;
+        self.static_servers.insert(
+            new_interface.id.clone(),
+            StaticServer::from_configs(
+                new_interface.clone(),
+                self.config.ui_bundle_by_id(&new_interface.bundle).unwrap(),
+                None,
+            ),
+        );
+        Ok(())
+    }
+
+    fn remove_ui_interface(&mut self, id: &String) -> Result<(), HolochainError> {
+        let to_stop = self
+            .config
+            .clone()
+            .ui_interfaces
+            .into_iter()
+            .filter(|ui_interface| ui_interface.id == *id);
+
+        for ui_interface in to_stop {
+            let _ = self.stop_ui_interface(&ui_interface.id);
+        }
+
+        let mut new_config = self.config.clone();
+        new_config.ui_interfaces = new_config
+            .ui_interfaces
+            .into_iter()
+            .filter(|ui_interface| ui_interface.id != *id)
+            .collect();
+        new_config.check_consistency()?;
+        self.config = new_config;
+        self.save_config()?;
+
+        self.static_servers
+            .remove(id)
+            .ok_or(HolochainError::ErrorGeneric(
+                "Could not remove server".into(),
+            ))?;
+
+        Ok(())
+    }
+
+    fn start_ui_interface(&mut self, id: &String) -> Result<(), HolochainInstanceError> {
+        let server = self.static_servers.get_mut(id)?;
+        notify(format!("Starting UI interface \"{}\"...", id));
+        server.start()
+    }
+
+    fn stop_ui_interface(&mut self, id: &String) -> Result<(), HolochainInstanceError> {
+        let server = self.static_servers.get_mut(id)?;
+        notify(format!("Stopping UI interface \"{}\"...", id));
+        server.stop()
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use super::*;
+    use container::admin::tests::*;
+    use std::{fs::File, io::Read};
+
+    #[test]
+    fn test_install_ui_bundle_from_file() {
+        let mut container = create_test_container("test_install_ui_bundle_from_file", 3000);
+        let bundle_path = PathBuf::from(".");
+        assert_eq!(
+            container.install_ui_bundle_from_file(bundle_path, &"test-bundle-id".to_string()),
+            Ok(())
+        );
```

Comment:
This test is not testing if the installed UI bundle is actually there after running the call. Just a smoke test?

Reply:
Actually at this stage the UI dir isn't copied at all just served from wherever it already is. I think we do actually want the copy option though so I will add that in